### PR TITLE
feat: add `created_at` and `updated_at` to wallets and accounts

### PR DESCRIPTION
**Pull request by dni**
the title says it all :)

### CODE REVIEW

File: lnbits/core/migrations.py [hunk 1]
```diff
@@ -404,3 +404,38 @@ async def m016_add_username_column_to_accounts(db):
...
+            ALTER TABLE accounts
+            ADD COLUMN created_at TIMESTAMP NOT NULL DEFAULT {db.timestamp_now}
+        """
+        )
+        await db.execute(
+            f"""
+            ALTER TABLE accounts
+            ADD COLUMN updated_at TIMESTAMP NOT NULL DEFAULT {db.timestamp_now}
+        """
+        )
```
Comment (motorina0) [inline lnbits/core/migrations.py:425]: nitpick: add the two columns at once
```suggestion
        await db.execute(
            f"""
            ALTER TABLE accounts
            ADD COLUMN created_at TIMESTAMP NOT NULL DEFAULT {db.timestamp_now},
            ADD COLUMN updated_at TIMESTAMP NOT NULL DEFAULT {db.timestamp_now}
            """
        )
```

File: lnbits/core/crud.py [hunk 2]
```diff
@@ -118,10 +118,16 @@ async def update_account(
 
     await db.execute(
         """
-            UPDATE accounts SET (username, email, extra) = (?, ?, ?)
+            UPDATE accounts SET (updated_at, username, email, extra) = (?, ?, ?, ?)
```
Comment (motorina0) [inline lnbits/core/crud.py:121]: nitpick: put `updated_at` last

File: lnbits/core/crud.py
Comment (motorina0) [inline lnbits/core/crud.py:1]: - select also columns `created_at` and `updated_at`:
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L142
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L198
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L209
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L227
- set `updated_at`:
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L182

Comment (motorina0) [inline lnbits/core/crud.py:1]: - set `updated_at`:
https://github.com/lnbits/lnbits/blob/742bf890797d932589058f0593059b1de0b299a1/lnbits/core/crud.py#L455

**Review by motorina0 — APPROVED** [review]
utACK
ps: tests fail

File: lnbits/core/crud.py [hunk 3]
```diff
@@ -173,9 +179,10 @@ async def update_user_password(data: UpdateUserPassword) -> Optional[User]:
     pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
 
     await db.execute(
-        "UPDATE accounts SET pass = ? WHERE id = ?",
+        "UPDATE accounts SET pass = ?, updated_at = ? WHERE id = ?",
         (
             pwd_context.hash(data.password),
+            db.timestamp_now,
```
Comment (jackstar12) [inline lnbits/core/crud.py:185]: does this actually work? `db.timestamp_now` just returns a string and not an actual timestamp

**Review by prusnak — CHANGES_REQUESTED** [review]
CI failed

File: lnbits/core/crud.py (hunk 3 above)
Comment (dni) [reply lnbits/core/crud.py]: take a look again pls :)

File: lnbits/core/migrations.py [hunk 4]
```diff
@@ -404,3 +404,43 @@ async def m016_add_username_column_to_accounts(db):
...
     except OperationalError:
         pass
+
+
+async def m017_add_timestamp_columns_to_accounts_and_wallets(db):
+    """
+    Adds created_at and updated_at column to accounts and wallets.
+    """
+    try:
+        if db.is_postgres:
```
Comment (motorina0) [inline lnbits/core/migrations.py:414]: This sets a bad precedent. All db type decisions should be encapsulated in the `db.py` file.

How about adding this in `db.py`:

```py
    @property
    def timestamp_column_default(self) -> str:
        if self.type in {POSTGRES, COCKROACH}:
            return self.timestamp_now
        return "NULL"
```

Then also allow `NULL` for PostgreSQL (the data model must reflect his anyhow)

File: lnbits/core/migrations.py [hunk 5]
```diff
@@ -404,3 +404,43 @@ async def m016_add_username_column_to_accounts(db):
...
+            )
+            await db.execute(
+                "ALTER TABLE accounts ADD COLUMN updated_at TIMESTAMP DEFAULT NULL"
+            )
+            await db.execute(
+                "ALTER TABLE wallets ADD COLUMN created_at TIMESTAMP DEFAULT NULL"
+            )
+            await db.execute(
+                "ALTER TABLE wallets ADD COLUMN updated_at TIMESTAMP DEFAULT NULL"
+            )
```
Comment (motorina0) [inline lnbits/core/migrations.py:442]: Would this work?
```suggestion
            await db.execute(
                f"ALTER TABLE accounts ADD COLUMN created_at TIMESTAMP DEFAULT {db.timestamp_column_default}"
            )
            await db.execute(
                f"ALTER TABLE accounts ADD COLUMN updated_at TIMESTAMP DEFAULT {db.timestamp_column_default}"
            )
            await db.execute(
                f"ALTER TABLE wallets ADD COLUMN created_at TIMESTAMP DEFAULT {db.timestamp_column_default}"
            )
            await db.execute(
                f"ALTER TABLE wallets ADD COLUMN updated_at TIMESTAMP DEFAULT {db.timestamp_column_default}"
            )
```

File: lnbits/db.py [hunk 6]
```diff
@@ -83,33 +83,41 @@ def datetime_to_timestamp(self, date: datetime.datetime):
             return time.mktime(date.timetuple())
         return "<nothing>"
 
+    @property
+    def is_postgres(self) -> bool:
```
Comment (motorina0) [inline lnbits/db.py:87]: this information should never be leaked outside this file

**Review by prusnak — CHANGES_REQUESTED** [review]
this shouldn't be merged yet - there are still unaddressed comments about leaking abstraction from `db.py` - changing to draft state

File: lnbits/db.py (hunk 6 above)
Comment (prusnak) [reply lnbits/db.py:87]: also the function is not doing what it is announcing, cockroach is not postgres, NACK

**Review by motorina0 — APPROVED** [review]
tACK

Nitpick: 
 - all existing wallets and accounts will have `created_at` the date of the migration
 - for a backwards valid data set a script can be created to set `created_at` at the first payment date

**Review by prusnak — APPROVED** [review]

**Review by motorina0 — APPROVED** [review]

**Pull request merged**